src/run.rs

```diff
@@ -0,0 +1,188 @@
+            1 => println!("1 file"),
+            n => println!("{} files", n),
+        }
+    }
+
+    match errors {
+        0 => Ok(()),
+        1 => Err(String::from("1 failure")),
+        n => Err(format!("{} failures", n)),
+    }
+}
+
+/// Iterate over all of the files passed as arguments, recursively iterating through directories
+fn iterate_files(files: Vec<String>) -> impl Iterator<Item = PathBuf> {
+    files
+        .into_iter()
+        .flat_map(WalkDir::new)
+        .filter(|f| match f {
+            Ok(d) => !is_hidden(d) && !is_directory(d),
+            Err(e) => {
+                println!("Unable to read file: {}", e);
+                false
+            }
+        })
+        .map(|f| {
+            f.expect("This should not happen: we have already filtered out the errors")
+                .into_path()
+        })
+}
+fn is_hidden(d: &DirEntry) -> bool {
+    d.file_name().to_str().map_or(false, |s| s.starts_with("."))
+}
+fn is_directory(d: &DirEntry) -> bool {
+    d.file_type().is_dir()
+}
+
+/// Run all functions in a file that are succeeded by "run:" comments
+fn run_single_file(path: &PathBuf) -> Result<(), String> {
+    let file_contents = read_to_string(&path).map_err(|e| e.to_string())?;
+    run_file_contents(file_contents)
+}
+
+/// Main body of `run_single_file` separated for testing
+fn run_file_contents(file_contents: String) -> Result<(), String> {
+    let test_file = parse_test(&file_contents, None, None).map_err(|e| e.to_string())?;
+    let isa = create_target_isa(test_file.isa_spec)?;
+    for (func, Details { comments, .. }) in test_file.functions {
+        if comments.iter().filter(|c| c.text.contains("run")).count() > 0 {
+            run_single_function(func, isa.as_ref())?
+        }
+    }
+    Ok(())
+}
+
+/// Build an ISA based on the current machine running this code (the host)
+fn create_target_isa(isa_spec: IsaSpec) -> Result<Box<dyn TargetIsa>, String> {
+    if let IsaSpec::None(flags) = isa_spec {
+        // build an ISA for the current machine
+        let builder = host_isa_builder()?;
+        Ok(builder.finish(flags))
+    } else {
+        Err(String::from("A target ISA was specified in the file but should not have been--only the host ISA can be used for running CLIF files"))?
+    }
+}
+
+/// Compile and execute a single function, expecting a boolean to be returned; a 'true' value is
+/// interpreted as a successful test execution and mapped to Ok whereas a 'false' value is
+/// interpreted as a failed test and mapped to Err.
+fn run_single_function(func: Function, isa: &dyn TargetIsa) -> Result<(), String> {
+    if !(func.signature.params.is_empty()
+        && func.signature.returns.len() == 1
+        && func.signature.returns.first().unwrap().value_type.is_bool())
+    {
+        return Err(String::from(
+            "Functions must have a signature like: () -> boolean",
+        ));
+    }
+    // TODO observed segfaults with return types of b64
```

Comment:
nit: remaining TODO in code; either we can open a follow-up issue, or it should be handled :)

Reply:
Ok, opened https://github.com/CraneStation/cranelift/issues/911; may need some help figuring this one out.